subsys/bluetooth/host/ecc.c

```diff
@@ -71,23 +104,143 @@ bool bt_pub_key_is_valid(const uint8_t key[BT_PUB_KEY_LEN])
 	return false;
 }
 
-int bt_pub_key_gen(struct bt_pub_key_cb *new_cb)
+static void set_key_attributes(psa_key_attributes_t *attr)
+{
+	psa_set_key_type(attr, PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1));
+	psa_set_key_bits(attr, 256);
+	psa_set_key_usage_flags(attr, PSA_KEY_USAGE_EXPORT | PSA_KEY_USAGE_DERIVE);
+	psa_set_key_algorithm(attr, PSA_ALG_ECDH);
+}
+
+static void generate_pub_key(struct k_work *work)
 {
+	psa_key_attributes_t attr = PSA_KEY_ATTRIBUTES_INIT;
 	struct bt_pub_key_cb *cb;
+	psa_key_id_t key_id;
+	uint8_t tmp_pub_key_buf[BT_PUB_KEY_LEN + 1];
+	size_t tmp_len;
+	int err;
+
+	set_key_attributes(&attr);
+
+	if (psa_generate_key(&attr, &key_id) != PSA_SUCCESS) {
+		LOG_ERR("Failed to generate ECC key");
+		err = BT_HCI_ERR_UNSPECIFIED;
+		goto done;
+	}
+
+	if (psa_export_public_key(key_id, tmp_pub_key_buf, sizeof(tmp_pub_key_buf),
+				  &tmp_len) != PSA_SUCCESS) {
+		LOG_ERR("Failed to export ECC public key");
+		err = BT_HCI_ERR_UNSPECIFIED;
+		goto done;
+	}
+	/* secp256r1 PSA exported public key has an extra 0x04 predefined byte at
+	 * the beginning of the buffer which is not part of the coordinate so
+	 * we remove that.
+	 */
+	memcpy(ecc.public_key_be, &tmp_pub_key_buf[1], BT_PUB_KEY_LEN);
+
+	if (psa_export_key(key_id, ecc.private_key_be, BT_PRIV_KEY_LEN,
+			&tmp_len) != PSA_SUCCESS) {
+		LOG_ERR("Failed to export ECC private key");
+		err = BT_HCI_ERR_UNSPECIFIED;
+		goto done;
+	}
+
+	if (psa_destroy_key(key_id) != PSA_SUCCESS) {
+		LOG_ERR("Failed to destroy ECC key ID");
+		err = BT_HCI_ERR_UNSPECIFIED;
+		goto done;
+	}
+
+	sys_memcpy_swap(pub_key, ecc.public_key_be, BT_PUB_KEY_COORD_LEN);
+	sys_memcpy_swap(&pub_key[BT_PUB_KEY_COORD_LEN],
+			&ecc.public_key_be[BT_PUB_KEY_COORD_LEN], BT_PUB_KEY_COORD_LEN);
+	err = 0;
+
+done:
+	atomic_clear_bit(flags, PENDING_PUB_KEY);
+	atomic_set_bit(bt_dev.flags, BT_DEV_HAS_PUB_KEY);
+
+	k_sched_lock();
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&pub_key_cb_slist, cb, node) {
+		if (cb->func) {
+			cb->func(err ? NULL : pub_key);
+		}
+	}
+
+	sys_slist_init(&pub_key_cb_slist);
+
+	k_sched_unlock();
```

Comment:
Isn't it risky to do a `k_sched_lock` before calling a list of callbacks? If any of them blocks?

Reply:
https://github.com/zephyrproject-rtos/zephyr/blob/176676d81fdc44a62fe6f5ffe857d0f521ee1f24/include/zephyr/kernel.h#L1231-L1232

Reply:
I added code comments, explaining what the call means in practice